akka-typed/src/main/scala/akka/typed/persistence/internal/PersistentActorImpl.scala

```diff
@@ -110,15 +110,17 @@ import akka.typed.internal.adapter.ActorRefAdapter
     case Persist(event) ⇒
       // apply the event before persist so that validation exception is handled before persisting
       // the invalid event, in case such validation is implemented in the event handler.
+      // also, we ensure, before persisting, that there is an event handler for each single event
       state = applyEvent(state, event)
       persist(event) { _ ⇒
         sideEffects.foreach(applySideEffect)
       }
     case PersistAll(events) ⇒
       // apply the event before persist so that validation exception is handled before persisting
       // the invalid event, in case such validation is implemented in the event handler.
+      // also, we ensure, before persisting, that there is an event handler for each single event
       state = events.foldLeft(state)(applyEvent)
-      persistAll(scala.collection.immutable.Seq(events)) { _ ⇒
+      persistAll(events) { _ ⇒
         sideEffects.foreach(applySideEffect)
```

Comment:
I think there is a bug here (not introduced by you). The callback is invoked for each event. It should be once, for the last event, as in https://github.com/lightbend/akka-commercial-addons/blob/master/akka-persistence-multi-dc/src/main/scala/akka/persistence/multidc/internal/ReplicatedEntityActor.scala#L443

Could you fix that at the same time? Could also be done in separate PR.

Reply:
and it should also handle the case when `events.isEmpty` to `applySideEffect` immediately.

Reply:
Yes, I saw this. I was working on that on a separated PR. Glad to add to this one. 

There is also another bug, which is the side-effects callbacks are not being added together. We only apply the first one. I have a test and a fix for that.